lib/std/io_uring.zig

```diff
@@ -0,0 +1,841 @@
+        errdefer os.munmap(mmap_sqes);
+        assert(mmap_sqes.len == size_sqes);
+
+        const array = @ptrCast([*]u32, @alignCast(@alignOf(u32), &mmap[p.sq_off.array]));
+        const sqes = @ptrCast([*]io_uring_sqe, @alignCast(@alignOf(io_uring_sqe), &mmap_sqes[0]));
+        // We expect the kernel copies p.sq_entries to the u32 pointed to by p.sq_off.ring_entries,
+        // see https://github.com/torvalds/linux/blob/v5.8/fs/io_uring.c#L7843-L7844.
+        assert(
+            p.sq_entries ==
+            @ptrCast(*u32, @alignCast(@alignOf(u32), &mmap[p.sq_off.ring_entries])).*
+        );
+        return SubmissionQueue {
+            .head = @ptrCast(*u32, @alignCast(@alignOf(u32), &mmap[p.sq_off.head])),
+            .tail = @ptrCast(*u32, @alignCast(@alignOf(u32), &mmap[p.sq_off.tail])),
+            .mask = @ptrCast(*u32, @alignCast(@alignOf(u32), &mmap[p.sq_off.ring_mask])),
+            .flags = @ptrCast(*u32, @alignCast(@alignOf(u32), &mmap[p.sq_off.flags])),
+            .dropped = @ptrCast(*u32, @alignCast(@alignOf(u32), &mmap[p.sq_off.dropped])),
+            .array = array[0..p.sq_entries],
+            .sqes = sqes[0..p.sq_entries],
+            .mmap = mmap,
+            .mmap_sqes = mmap_sqes
+        };
+    }
+
+    pub fn deinit(self: *SubmissionQueue) void {
+        os.munmap(self.mmap_sqes);
+        os.munmap(self.mmap);
+    }
+};
+
+pub const CompletionQueue = struct {
+    head: *u32,
+    tail: *u32,
+    mask: *u32,
+    overflow: *u32,
+    cqes: []io_uring_cqe,
+
+    pub fn init(fd: i32, p: io_uring_params, sq: SubmissionQueue) !CompletionQueue {
+        assert(fd >= 0);
+        assert((p.features & linux.IORING_FEAT_SINGLE_MMAP) > 0);
+        const mmap = sq.mmap;
+        const cqes = @ptrCast(
+            [*]io_uring_cqe,
+            @alignCast(@alignOf(io_uring_cqe), &mmap[p.cq_off.cqes])
+        );
+        assert(
+            p.cq_entries ==
+            @ptrCast(*u32, @alignCast(@alignOf(u32), &mmap[p.cq_off.ring_entries])).*
+        );
+        return CompletionQueue {
+            .head = @ptrCast(*u32, @alignCast(@alignOf(u32), &mmap[p.cq_off.head])),
+            .tail = @ptrCast(*u32, @alignCast(@alignOf(u32), &mmap[p.cq_off.tail])),
+            .mask = @ptrCast(*u32, @alignCast(@alignOf(u32), &mmap[p.cq_off.ring_mask])),
+            .overflow = @ptrCast(*u32, @alignCast(@alignOf(u32), &mmap[p.cq_off.overflow])),
+            .cqes = cqes[0..p.cq_entries]
+        };
+    }
+
+    pub fn deinit(self: *CompletionQueue) void {
+        // A no-op since we now share the mmap with the submission queue.
+        // Here for symmetry with the submission queue, and for any future feature support.
+    }
+};
+
+inline fn check_errno(res: usize) !void {
+    const errno = linux.getErrno(res);
+    if (errno != 0) return os.unexpectedErrno(errno);
+}
+
+const minimum = std.Target.current.os.isAtLeast;
+pub const can_single_mmap = comptime minimum(.linux, .{ .major = 5, .minor = 4 }) == true;
```

Comment:
This is just a bound: if the minimum version is lower, then you need to check at runtime.

Reply:
Thanks, is the runtime check the same `std.Target.current.os.isAtLeast()` call... but without `comptime`? Or do we need to use something else?

Reply:
No `Target.current` gets it from `builtin`, you have to do a syscall or something to find os version at runtime

Reply:
I think right now `Target.current` is confusing. It should probably made a bit more clear like `Target.current_build`.

Reply:
Here `can_single_mmap` is implicitly comptime, you dont have to use the keyword.